### PR TITLE
JDK-8315464: Uncouple AllClassesIndexWriter from IndexBuilder

### PR DESCRIPTION
**Pull request by hns**
My original intention with this cleanup was to keep using `[Html]IndexBuilder' for both the "All Classes" and the main index, but just improve separation of code and naming of classes. However, I then discovered two things:

 - The "All Classes" page is so simple to build that it doesn't justify using a complex builder class. In fact, moving the necessary functionality from `IndexBuilder` to `AllClassesIndexWriter` just added about a dozen lines of code to the latter, while making `IndexBuilder` much simpler by removing its dual purpose.
 - The separation of code between `toolkit.util.IndexBuilder` and `formats.html.HtmlIndexBuilder`, although seemingly random at first sight, actually reflects the usage of functionality provided by the respective packages. Although with a single output format the separation between these packages is not as important as it used to be, it is still nice to keep things separated.

Based on these two findings, I decided to pull the "All Classes" code into "AllClassesIndexWriter", but maintain the implementation of `[Html]IndexBuilder` across the two classes/packages. Although the base `IndexBuilder` class is technically able to function on its own without the HTML-specific enhancements, I made the class abstract in order to point out that the HTML enhancements in `HtmlIndexBuilder` are actually essential for our us.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315464](https://bugs.openjdk.org/browse/JDK-8315464): Uncouple AllClassesIndexWriter from IndexBuilder (**Task** - P4)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15746/head:pull/15746` \
`$ git checkout pull/15746`

Update a local copy of the PR: \
`$ git checkout pull/15746` \
`$ git pull https://git.openjdk.org/jdk.git pull/15746/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15746`

View PR using the GUI difftool: \
`$ git pr show -t 15746`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15746.diff">https://git.openjdk.org/jdk/pull/15746.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15746#issuecomment-1719773603)

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AllClassesIndexWriter.java [hunk 1]
```diff
@@ -115,6 +104,24 @@ protected void addContents(Content target) {
         }
     }
 
+    private Set<TypeElement> getTypeElements()  {
```
Comment (turbanoff) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AllClassesIndexWriter.java:107]:
```suggestion
    private Set<TypeElement> getTypeElements() {
```

**Review by jonathan-gibbons — APPROVED** [review]
Approved.

I assume all tests pass, and that JDK API docs are not affected.

**Comment by hns** [comment]
Thanks for reviewing!

> I assume all tests pass, and that JDK API docs are not affected.

Yes to both.

**Pull request closed without merging**